internal/match/regexp.go

```diff
@@ -88,3 +89,123 @@ func (re *LRE) match(text string) bool {
 func (re *LRE) compile() {
 	re.dfa = reCompileDFA(re.syntax.compile(nil, 0))
 }
+
+// A MultiLRE matches multiple LREs simultaneously against a text.
+// It is more efficient than matching each LRE in sequence against the text.
+type MultiLRE struct {
+	dict *Dict
+
+	// groups maps a two-word input text phrase
+	// to the group of LREs that might match at that location.
+	groups map[phrase]*reGroup
+}
+
+// A phrase is a phrase of up to two words.
+// The zero-word phrase is phrase{NoWord, NoWord}.
+// A single-word phrase w is phrase{w, NoWord}.
+type phrase [2]WordID
+
+// A reGroup a group of LREs that can possibly match text
+// beginning with a given phrase.
+type reGroup struct {
+	list []*LRE
+	id   []int
+	dfa  reDFA
+}
+
+// NewMultiLRE returns a MultiLRE looking for the given LREs.
+// All the LREs must have been parsed using the same Dict;
+// if not, NewMultiLRE panics.
+func NewMultiLRE(list []*LRE) (_ *MultiLRE, err error) {
+	if len(list) == 0 {
+		return &MultiLRE{}, nil
+	}
+
+	dict := list[0].dict
+	for _, sub := range list[1:] {
+		if sub.dict != dict {
+			panic("MultiRE: LREs parsed with different Dicts")
+		}
+	}
+
+	groups := make(map[phrase]*reGroup)
+	for subID, sub := range list {
+		phrases := sub.syntax.leadingPhrases()
+		if len(phrases) == 0 {
+			return nil, fmt.Errorf("%s: no leading phrases", sub.File())
+		}
+		for _, p := range phrases {
+			if p[1] == BadWord {
+				return nil, fmt.Errorf("%s: no 2-word leading phrases", sub.File())
+			}
+			g := groups[p]
+			if g == nil {
+				g = new(reGroup)
+				groups[p] = g
+			}
+			g.list = append(g.list, sub)
+			g.id = append(g.id, subID)
+		}
+	}
+
+	for _, g := range groups {
```

Comment:
It's true that the code could make a single DFA of all the LREs and then do:

	for i := 0; i < len(words); i++ {
		tryMatch(words[i:])
	}

The overhead of starting up a match is non-trivial, though, and if you know it won't pay off, because even the first word is wrong, then you could do a little preprocessing and avoid calling tryMatch sometimes:

	start := make(map[WordID]bool)
	for _, lre := range list {
		for _, w := range possibleFirstWords(lre) {
			start[lre] = true
		}
	}

	for i := 0; i < len(words); i++ {
		if start[words[i]] {
			tryMatch(words[i:])
		}
	}

It turns out to be the case that recording pairs of words (aka phrases) is even more effective for licenses—that is, the benefit outweighs the cost of the more expensive lookup—so we can do:

	start := make(map[phrase]bool)
	for _, lre := range list {
		for _, p := range possibleFirstPhrases(lre) {
			start[p] = true
		}
	}

	for i := 0; i < len(words)-1; i++ {
		if start[phrase{words[i], words[i+1]}] {
			tryMatch(words[i:])
		}
	}

That would work fine with one big DFA. But the DFA's transition table for each state is a list of (word ID, next state) pairs scanned in linear order. For licenses, nearly all the time you're looking for one specific word, or at worst a few different possible next words, and so the linear scan is fine, perhaps even faster than more complex things like binary search.

The one exception is at the start of a match, when you don't know anything about which licenses you are looking for. That initial state would have a transition table with one entry for every possible first word of a license, which in the current SPDX set would mean 179 entries. We could do a binary search for big entries like that and get it down to 8 probes. Or we could avoid the problem and make a different DFA for each starting phrase, which is what the code does.

That said, I made this decision based on things like the MIT and BSD licenses, where there's a single obvious start (Permission is hereby granted... or Redistribution and use ...). A bunch of the much larger licenses have some optional early paragraphs, which are important to recognize if there, because they affect the calculation of how much of the file got matched, but also important to be able to do without. These licenses have multiple starts, and breaking up the DFA by start phrase has the effect of putting those licenses into multiple DFAs, which duplicates their DFA footprint. I did the experiment, and using just the one DFA takes up only 84 MB instead of 106 MB. That's a nice win.

And I measured whether match speed dropped, and it basically didn't, even if I only did it for the large (179-entry) initial state transition table. Linear scans are just much more friendly to modern processors I guess.

Anyway, a very long-winded way of saying: Done. Thanks.